### ISSUE TITLE
Request for guidance building a jupyterlab extension that makes heavy use of ipywidgets

### ISSUE THREAD

**Issue by SimonBiggs** [jupyter-widgets/ipywidgets#1797]
Hi @jasongrout ,

I am currently building the following jupyterlab extension:
https://github.com/SimonBiggs/jupyterlab-form

The gist of it is users can write a python enabled form in markdown. The focus is on having very small amounts of code so as not to scare off non-python users. The results of the form inputs/outputs need to be saved to json so that another extension, written later, can be trend the data etc. The markdown itself ends up being a form template that can be easily updated as needed.

I already have some basic interactivity, however I believe that what I am building would be significantly benefited by trying to leverage the work you've done with ipywidgets.

At the moment I have just two input types, number and string. Essentially when those inputs change then the python code in the same section is run. There is currently no feedback from the kernel to update those variables if their representation changes in the kernel.

Do you have recommendations of how I might be able to go about including some of the ipywidgets interfaces into my extension? Ideally the method I would use would be sufficiently decoupled so that other ipywidgets could be included at a later stage. It would be nice if I provided a simple custom tag for the major ipywidgets, but then also provided a general ipywidgets tag allowing any other ipywidget to be inserted.

What are your thoughts? Is trying to build atop of ipywidgets worthwhile, or would it be the wrong way to achieve what I am after? From what I can tell building ipywidgets into the forms achieves what I am after perfectly.

**Comment by jasongrout** [comment]
I'm not quite understanding how your system works. Can you perhaps give an example markdown file describing a form and a screenshot of the rendered form, and explain exactly how it works? It sounds like a kernel is associated with the markdown document? Or maybe just a rendered form from the markdown document?

Widgets, by their nature, can be rendered into any html context, and communicate back and forth to a kernel. If you are rendering a form to html, likely you can render widgets if you write a widget manager to do so (you could maybe base such a widget manager on the html widget manager?). Again, I'm trying to understand better what you are proposing and how your system works.

**Comment by SimonBiggs** [comment]
Hi @jasongrout,

Perhaps the easiest way is to just show you? 

https://github.com/SimonBiggs/jupyterlab-form/blob/master/README.md

If you install that jupyterlab extension and then from the launcher panel press "form", a demo form will be created. The explanation text is a bit out of date because since I wrote that text I also made a table variable. But nevertheless it gives a good idea of what's going on.

If you also attach a console to the same kernel you'll be able to see the calls I am making to sync up the variables.

Since I wrote this message I have actually managed to implement a variable sync. I'm keen for direction and feedback though.

I'm also struggling how to actually save the results of the document... The results file will be json format and will contain a copy of the template from the time of filling it out, but it should also contain a copy of all outputs and inputs that were provided when it was saved.
